Source/JavaScriptCore/runtime/BufferMemoryHandle.h

```diff
@@ -0,0 +1,177 @@
+#include <wtf/StdSet.h>
+#include <wtf/Vector.h>
+#include <wtf/WeakPtr.h>
+
+namespace WTF {
+class PrintStream;
+}
+
+namespace JSC {
+
+class LLIntOffsetsExtractor;
+
+enum class GrowFailReason : uint8_t {
+    InvalidDelta,
+    InvalidGrowSize,
+    WouldExceedMaximum,
+    OutOfMemory,
+    GrowSharedUnavailable,
+};
+
+struct BufferMemoryResult {
+    enum Kind {
+        Success,
+        SuccessAndNotifyMemoryPressure,
+        SyncTryToReclaimMemory
+    };
+
+    static ASCIILiteral toString(Kind);
+
+    BufferMemoryResult() { }
+
+    BufferMemoryResult(void* basePtr, Kind kind)
+        : basePtr(basePtr)
+        , kind(kind)
+    {
+    }
+
+    void dump(PrintStream&) const;
+
+    void* basePtr;
+    Kind kind;
+};
+
+class BufferMemoryManager {
+    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_NONCOPYABLE(BufferMemoryManager);
+public:
+    friend class LazyNeverDestroyed<BufferMemoryManager>;
+
+#if ENABLE(WEBASSEMBLY_SIGNALING_MEMORY)
+    BufferMemoryResult tryAllocateFastMemory();
+    void freeFastMemory(void* basePtr);
+#endif
+
+    BufferMemoryResult tryAllocateGrowableBoundsCheckingMemory(size_t mappedCapacity);
+
+    void freeGrowableBoundsCheckingMemory(void* basePtr, size_t mappedCapacity);
+
+    bool isInGrowableOrFastMemory(void* address);
+
+    // We allow people to "commit" more wasm memory than there is on the system since most of the time
+    // people don't actually write to most of that memory. There is some chance that this gets us
+    // JetSammed but that's possible anyway.
+    inline size_t memoryLimit() const
+    {
+        if (productOverflows<size_t>(ramSize(),  3))
+            return std::numeric_limits<size_t>::max();
+        return ramSize() * 3;
+    }
+
+    // FIXME: Ideally, bmalloc would have this kind of mechanism. Then, we would just forward to that
+    // mechanism here.
+    BufferMemoryResult::Kind tryAllocatePhysicalBytes(size_t bytes);
+
+    void freePhysicalBytes(size_t bytes);
+
+    void dump(PrintStream& out) const;
+
+    static BufferMemoryManager& singleton();
+
+private:
+    BufferMemoryManager() = default;
+
+    Lock m_lock;
+#if ENABLE(WEBASSEMBLY_SIGNALING_MEMORY)
+    unsigned m_maxFastMemoryCount { Options::maxNumWebAssemblyFastMemories() };
+    Vector<void*> m_fastMemories;
+#endif
+    StdSet<std::pair<uintptr_t, size_t>> m_growableBoundsCheckingMemories;
+    size_t m_physicalBytes { 0 };
+};
+
+class BufferMemoryHandle final : public ThreadSafeRefCounted<BufferMemoryHandle> {
+    WTF_MAKE_NONCOPYABLE(BufferMemoryHandle);
+    WTF_MAKE_FAST_ALLOCATED;
+    friend LLIntOffsetsExtractor;
+public:
+    BufferMemoryHandle(void*, size_t size, size_t mappedCapacity, PageCount initial, PageCount maximum, MemorySharingMode, MemoryMode);
+    JS_EXPORT_PRIVATE ~BufferMemoryHandle();
+
+    void* memory() const;
+    size_t size(std::memory_order order = std::memory_order_seq_cst) const
+    {
+        if (m_sharingMode == MemorySharingMode::Default)
```

Comment:
Not sure since BufferMemoryHandle can be mainly used for shared or not. So not having any predicate here (LIKELY / UNLIKELY) since they can be neutral.